### ISSUE TITLE
Not all image references updated by linkerd install --registry

### ISSUE THREAD

**Issue by brendonthiede** [linkerd/linkerd2#5863]
### What is the issue?

When running `linkerd install --registry=<private registry>` not all image references are updated.

### How can it be reproduced?

I ran the follow to get a list of all images in the generated resources:

```bash
linkerd install --ignore-cluster --registry=private-cr/linkerd | grep 'image:' -A1 | grep -v 'image:$\|imagePullPolicy:\|--' | awk '{print $2}' | sort -u
```

which produces:

```

cr.l5d.io/linkerd/grafana:stable-2.9.4
private-cr/linkerd/controller:stable-2.9.4
private-cr/linkerd/debug
private-cr/linkerd/proxy
private-cr/linkerd/proxy-init
private-cr/linkerd/proxy-init:v1.3.9
private-cr/linkerd/proxy:stable-2.9.4
private-cr/linkerd/web:stable-2.9.4
prom/prometheus:v2.19.3
```

My expectation is that the `cr.l5d.io/linkerd/grafana:stable-2.9.4` would be updated to `private-cr/linkerd/grafana:stable-2.9.4`. For my situation, the `prom/prometheus:v2.19.3` reference would need to be updated to `private-cr/prom/prometheus:v2.19.3`, but that follows a different pattern than the others (doesn't have /linkerd at the front), so that would need to be handled differently.

### Logs, error output, etc

I end up with pods in `ImagePullBackoff` unless I manipulate the generated resource definitions

#### `linkerd check` output

N/A

### Environment

- Kubernetes Version: 1.18.9
- Cluster Environment: OpenStack (manual binary installation)
- Host OS: Ubuntu 16.04.7 LTS
- Linkerd version: stable-2.9.4

### Possible solution

I'm not sure where this is coming from, but my current workaround is to use:

```bash
linkerd install --registry=<private-repo>/linkerd | sed -E 's/image: (cr.l5d.io\/|(prom\/))/image: <private-repo>\/\2/g' | kubectl apply -f -
```

This has things working for us as of now.

### Additional context

None

**Comment by alpeb** [comment]
Thanks for the report @brendonthiede :+1: 

I'm happy to report this has already been addressed in the latest edge and will be fixed for 2.10:
```console
$ linkerd install --ignore-cluster --registry=private-cr/linkerd | grep 'image:' -A1 | grep -v 'image:$\|imagePullPolicy:\|--' | awk '{print $2}' | sort -u
private-cr/linkerd/controller:edge-21.3.1
private-cr/linkerd/debug
private-cr/linkerd/proxy
private-cr/linkerd/proxy:edge-21.3.1
private-cr/linkerd/proxy-init
private-cr/linkerd/proxy-init:v1.3.9

$ linkerd viz install --skip-checks --set defaultRegistry=private-cr/linkerd --set prometheus.image.registry=private-cr/linkerd | grep 'image:' -A1 | grep -v 'image:$\|imagePullPolicy:\|--' | awk '{print $2}' | sort -u
private-cr/linkerd/grafana:edge-21.3.1
private-cr/linkerd/metrics-api:edge-21.3.1
private-cr/linkerd/prometheus:v2.19.3
private-cr/linkerd/tap:edge-21.3.1
private-cr/linkerd/web:edge-21.3.1
```
Note that Prometheus' registry doesn't honor the `defaultRegistry` config because it has to point to the `prom` registry, so it needs to be overridden explicitly as shown above.

**Comment by brendonthiede** [comment]
Awesome! Thanks!
